_posts/2017-04-12-managing-technical-debt.md

```diff
@@ -0,0 +1,83 @@
+7. **Create unit/integration tests infrastructure** in your application, so everybody in a team will be able to create and run the tests locally
+8. **Build Continuous Integration** to run tests on a regular bases and collect the metrics from your code.
+9. **Build one-step deployment process**. [Automatic deployment from CI](http://www.savvyclutch.com/devops/continuous-deployment-to-aws-ecs-and-circle-ci/) is the perfect solution. 
+10. **Use one code style**, and use tool on CI to check it.
+11. **Start [doing code reviews](http://www.savvyclutch.com/process/Make-Code-Review-Useful-Again/)**
+12. **Triage and move all technical issues/stories in you issue tracker application to the one list**, so you will be able to see whole picture of your real technical debt and have realistic plan for it. 
+13. **Create template for new issues/bugs/ stories in the issue tracker.** Without good descriptions in your issues they became an absolutely useless. This is necessary bureaucracy to save developers time and focus. 
+14. **Investigate most confusing and affective parts in application** and create a plan to how to make them better
+15. **Start creating tests for all newly created code**, and insist on them on code reviews (at least acceptance tests should cover the issue).
+16. **Follow the “The Boy Scouts Rule”**, create the tests for part you touch and refactor it. Try to spend on this at least 20% of time.  
+
+# LR: Details
+So, where is the technical debt most often accumulated? On a new project team usually:
+
+1. Fear to make the changes
+2. Has no completely understanding of the application domain logic 
+3. Skip unit/functional/acceptance/performance tests, and testing infrastructure at all
+4. Implement manual deployment strategy and infrastructure configuration
+5. Do not track requirements
+6. Configure Dev environment and onboarding manually
+7. Has no backup automation for data
+8. Has no data sampling for dev/QA purposes 
+9. Has a lot of “code smells” in source code, design by committee with a lot of hidden requirements due to lack of refactoring
+10. Has a lot of bugs and requests with the poor descriptions, divided between 3-4 different projects/boards/lists in the issue tracker 
+11. Has almost no process or procedures how to dealing with issues/features
+
+Let me describe each a little. 
+
+From my experience, the most destructive, hard to fix and influencing the speed of development thing is infrastructure. This including: application configuration management, application requirements management, dev environment, deployment process, testing infrastructure (local and CI) and commitment process. Without any of this the efficiency fall dramastically with no chance to become better. 
+
+Without configuration management it’s very hard to not introduce bugs on production because of difference between local machines and production machines and this can cause critical issues for application users. 
```

Comment:
Without configuration management it’s very [likely to introduce][hard not to introduce] bugs on production